cpp/src/arrow/csv/options.h

```diff
@@ -101,6 +102,8 @@ struct ARROW_EXPORT ConvertOptions {
   /// This option is ignored if `include_columns` is empty.
   bool include_missing_columns = false;
 
+  std::vector<std::unique_ptr<TimestampConverter>> timestamp_converters;
```

Comment:
It's not useful to include TimestampConverter here; just construct these from strings in `csv::TimestampConverter::Convert()`, the way [StringConverter<TimestampType> currently is](https://github.com/apache/arrow/pull/6631/files#diff-d72dc952edd36f9337da2cfe30253841R415)

```suggestion
  std::vector<std::string> timestamp_formats;
```

Reply:
As I told further, we need that interface for performance reasons, so we need to decide if we want to keep it before.

Reply:
I think it makes sense to have user-defined converters, so I'm fine with this interface

Reply:
```suggestion
  std::vector<std::shared_ptr<TimestampConverter>> timestamp_converters;
```
since this ConvertOptions must be copyable